src/metric-engine/src/engine/read.rs

```diff
@@ -0,0 +1,251 @@
+            state
+                .get_physical_region_id(logical_region_id)
+                .with_context(|| {
+                    error!("Trying to alter an nonexistent region {logical_region_id}");
+                    LogicalRegionNotFoundSnafu {
+                        region_id: logical_region_id,
+                    }
+                })?
+        };
+        let data_region_id = utils::to_data_region_id(physical_region_id);
+        let request = self
+            .transform_request(physical_region_id, logical_region_id, request)
+            .await?;
+        self.mito
+            .handle_query(data_region_id, request)
+            .await
+            .context(MitoReadOperationSnafu)
+    }
+
+    /// Transform the [ScanRequest] from logical region to physical data region.
+    async fn transform_request(
+        &self,
+        physical_region_id: RegionId,
+        logical_region_id: RegionId,
+        mut request: ScanRequest,
+    ) -> Result<ScanRequest> {
+        // transform projection
+        let physical_projection = if let Some(projection) = &request.projection {
+            self.transform_projection(physical_region_id, logical_region_id, projection)
+                .await?
+        } else {
+            self.default_projection(physical_region_id, logical_region_id)
+                .await?
+        };
+        request.projection = Some(physical_projection);
+
+        // add table filter
+        request
+            .filters
+            .push(self.table_id_filter(logical_region_id));
+
+        Ok(request)
+    }
+
+    /// Generate a filter on the table id column.
+    fn table_id_filter(&self, logical_region_id: RegionId) -> Expr {
+        logical_expr::col(DATA_SCHEMA_TABLE_ID_COLUMN_NAME)
+            .eq(logical_expr::lit(logical_region_id.table_id()))
+            .into()
+    }
+
+    /// Transform the projection from logical region to physical region.
+    ///
+    /// This method will not preserve internal columns.
+    pub async fn transform_projection(
+        &self,
+        physical_region_id: RegionId,
+        logical_region_id: RegionId,
+        origin_projection: &[usize],
+    ) -> Result<Vec<usize>> {
+        // project on logical columns
+        let logical_columns = self
+            .load_logical_columns(physical_region_id, logical_region_id)
+            .await?;
+
+        // generate physical projection
+        let mut physical_projection = Vec::with_capacity(origin_projection.len());
+        let data_region_id = utils::to_data_region_id(physical_region_id);
+        let physical_metadata = self
+            .mito
+            .get_metadata(data_region_id)
+            .await
+            .context(MitoReadOperationSnafu)?;
+        for logical_proj in origin_projection {
+            let column_id = logical_columns[*logical_proj].column_id;
+            // filter out internal columns
+            if ReservedColumnId::is_reserved(column_id) {
```

Comment:
Is it possible that a logical table contains a reserved column?

Reply:
Only when the column id overflows

Reply:
We can check column ids while creating table/adding columns if id overflowing is really a problem. I think we don't need to filter internal columns here. Otherwise we should also filter them in `default_projection()`.

Reply:
Default projection recovers columns from metadata, where the internal column is already filtered out.